feeder/cmd/pixel_bt_feeder/main.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// pixel_bt_feeder is an implementation of a witness feeder for the Pixel BT log.
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/internal/feeder"
+	"github.com/google/trillian-examples/serverless/config"
+	"golang.org/x/mod/sumdb/note"
+	"golang.org/x/mod/sumdb/tlog"
+
+	i_note "github.com/google/trillian-examples/internal/note"
+	wit_http "github.com/google/trillian-examples/witness/golang/client/http"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	// tileHeight is the tlog tile height.
+	// From: https://developers.devsite.corp.google.com/android/binary_transparency/tile
+	tileHeight = 1
+)
+
+var (
+	configFile = flag.String("config_file", "", "Path to feeder config file.")
+	timeout    = flag.Duration("timeout", 10*time.Second, "Maximum time to wait for witnesses to respond.")
+	interval   = flag.Duration("interval", time.Duration(0), "Interval between attempts to feed checkpoints. Default of 0 causes the tool to be a one-shot.")
+)
+
+// Config encapsulates the feeder config.
+type Config struct {
+	// Log defines the source log to feed from.
+	Log config.Log `yaml:"Log"`
+
+	// Witness is the configured witness.
+	Witness config.Witness `yaml:"Witness"`
+}
+
+func main() {
+	flag.Parse()
+
+	cfg, err := readConfig(*configFile)
+	if err != nil {
+		glog.Exitf("Failed to read config: %v", err)
+	}
+
+	u, err := url.Parse(cfg.Witness.URL)
+	if err != nil {
+		glog.Exitf("Failed to parse witness URL %q: %v", cfg.Witness.URL, err)
+	}
+	witness := wit_http.Witness{
+		URL:      u,
+		Verifier: mustCreateVerifier(i_note.Note, cfg.Witness.PublicKey),
+	}
+
+	ctx := context.Background()
+	if err := feedLog(ctx, cfg.Log, witness, *timeout, *interval); err != nil {
+		glog.Errorf("feedLog: %v", err)
+	}
+}
+
+// convertToNote converts a the PixelBT checkpoint to a valid Note.
+//
+// Hopefully we won't need this for too long, and PixelBT will update their checkpoint format to make it
+// fully compatible.
+func convertToNote(pixelCP, keyName string, keyHash uint32) ([]byte, error) {
```

Comment:
Please add a test for this, even if only for the happy path. It will make it easier to see what the input -> output transform is.

Reply:
yeah, thanks for keeping me honest here, done.